### PR TITLE
all: rename Edge methods Head and Tail to From and To

### LINKED ISSUE #51: all: semantics of head and tail are reversed cf. the rest of the world
I just noticed this today. From what I can see (this is not specified in the interface definition), graph.Edge.Head() returns the node at the non-pointy end of an edge, while graph.Edge.Tail() returns the node at the pointy end.

This is demonstrated, by the following:

```
package main

import (
	"fmt"

	"github.com/gonum/graph/concrete"
)

func main() {
	g := concrete.NewDirectedGraph()
	g.AddNode(concrete.Node(0))
	g.AddNode(concrete.Node(1))
	g.AddNode(concrete.Node(2))

	g.AddDirectedEdge(concrete.Edge{H: concrete.Node(0), T: concrete.Node(1)}, 0)
	g.AddDirectedEdge(concrete.Edge{H: concrete.Node(0), T: concrete.Node(2)}, 0)

	fmt.Println(g.Successors(concrete.Node(0)))
}
```

outputs:

```
[1 2]
```

This is the reverse of the normal interpretation.

>An arc e = (x, y) is considered to be directed from x to y; y is called the head and x is called the tail of the arc; y is said to be a direct successor of x, and x is said to be a direct predecessor of y.

I propose that we use simpler completely unambiguous terminology: `From` (conventionally tail) and `To` (conventionally head).

### PR DESCRIPTION
**Pull request by kortschak**
Also remove redundant search.johnsonGraphEdge type.

Fixes #51.

PTAL @vladimir-ch @gonum/developers

### CODE REVIEW

File: concrete/dense.go [hunk 1]
```diff
@@ -126,16 +126,16 @@ func (g *DenseGraph) Predecessors(n graph.Node) []graph.Node {
 }
 
 func (g *DenseGraph) Cost(e graph.Edge) float64 {
-	return g.adjacencyMatrix[e.Head().ID()*g.numNodes+e.Tail().ID()]
+	return g.adjacencyMatrix[e.From().ID()*g.numNodes+e.To().ID()]
 }
 
 // Sets the cost of an edge. If the cost is +Inf, it will remove the edge,
 // if directed is true, it will only remove the edge one way. If it's false it will change the cost
 // of the edge from succ to node as well.
```
Comment (vladimir-ch) [inline concrete/dense.go]: This could be updated, what is succ, what is node?

File: graph.go [hunk 2]
```diff
@@ -180,11 +180,11 @@ type MutableDirectedGraph interface {
 
 	// Like EdgeTo in DirectedGraph, AddDirectedEdge adds an edge FROM head TO tail.
```
Comment (vladimir-ch) [inline graph.go]: Update head, tail?

File: graph.go [hunk 3]
```diff
@@ -160,8 +160,8 @@ type MutableGraph interface {
 
 	// Like EdgeBetween in Graph, AddUndirectedEdge adds an edge between two nodes.
 	// If one or both nodes do not exist, the graph is expected to add them. However,
-	// if the nodes already exist it should NOT replace existing nodes with e.Head() or
-	// e.Tail(). Overwriting nodes should explicitly be done with another call to AddNode()
+	// if the nodes already exist it should NOT replace existing nodes with e.From() or
+	// e.To(). Overwriting nodes should explicitly be done with another call to AddNode()
```
Comment (vladimir-ch) [inline graph.go]: Dot at the end?

BTW, has there been a conclusion on if AddNode() should overwrite existing nodes or not?

File: graph.go [hunk 4]
```diff
@@ -180,11 +180,11 @@ type MutableDirectedGraph interface {
 
 	// Like EdgeTo in DirectedGraph, AddDirectedEdge adds an edge FROM head TO tail.
 	// If one or both nodes do not exist, the graph is expected to add them. However,
-	// if the nodes already exist it should NOT replace existing nodes with e.Head() or
-	// e.Tail(). Overwriting nodes should explicitly be done with another call to AddNode()
+	// if the nodes already exist it should NOT replace existing nodes with e.From() or
+	// e.To(). Overwriting nodes should explicitly be done with another call to AddNode()
```
Comment (vladimir-ch) [inline graph.go]: Dot?

**Comment by vladimir-ch** [comment]
LGTM except for the nitpicking.

File: concrete/dense.go (hunk 1 above)
Comment (kortschak) [inline concrete/dense.go]: I plan to do a lot more fix-ups, but I would like to keep the individual changes small and focused.

File: graph.go (hunk 3 above)
Comment (kortschak) [inline graph.go]: Again, fix-ups later.

Not yet, except in my mind. I think we should panic.

File: concrete/dense.go (hunk 1 above)
Comment (vladimir-ch) [inline concrete/dense.go]: Ok

File: graph.go (hunk 4 above)
Comment (kortschak) [inline graph.go]: As above.

File: graph.go (hunk 2 above)
Comment (kortschak) [inline graph.go]: Yes, and we need to sort out better language when the time comes.

**Pull request merged**